validation.go

```diff
@@ -0,0 +1,132 @@
+package merkle
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+)
+
+const MaxUint = ^uint(0)
+
+// ValidatePartialTree uses leafIndices, leaves and proof to calculate the merkle root of the tree and then compares it
+// to expectedRoot.
+func ValidatePartialTree(leafIndices []uint64, leaves, proof []Node, expectedRoot Node) (bool, error) {
+	v, err := newValidator(leafIndices, leaves, proof)
+	if err != nil {
+		return false, err
+	}
+	root := v.calcRoot(MaxUint)
+	return bytes.Equal(root, expectedRoot), nil
+}
+
+func newValidator(leafIndices []uint64, leaves, proof []Node) (validator, error) {
+	if len(leafIndices) != len(leaves) {
+		return validator{}, fmt.Errorf("number of leaves (%d) must equal number of indices (%d)", len(leaves), len(leafIndices))
+	}
+	if len(leaves) == 0 {
+		return validator{}, fmt.Errorf("at least one leaf is required for validation")
+	}
+	if len(leaves)+len(proof) == 1 {
+		return validator{}, fmt.Errorf("tree of size 1 not supported")
+	}
+	proofNodes := &proofIterator{proof}
+	leafIt := &leafIterator{leafIndices, leaves}
+
+	return validator{leafIt, proofNodes}, nil
+}
+
+type validator struct {
+	leaves     *leafIterator
+	proofNodes *proofIterator
+}
+
+func (v *validator) calcRoot(stopAtLayer uint) Node {
+	layer := uint(0)
+	idx, activeNode, err := v.leaves.next()
+	if err != nil {
+		panic(err) // this should never happen since we verify there are more leaves before calling calcRoot
+	}
+	var leftChild, rightChild, sibling Node
+	for {
+		if layer == stopAtLayer {
+			break
+		}
+		if v.shouldCalcSubtree(idx, layer) {
+			sibling = v.calcRoot(layer)
+		} else {
+			var err error
+			sibling, err = v.proofNodes.next()
+			if err == noMoreItems {
+				break
+			}
+		}
+		if leftSibling(idx, layer) {
+			leftChild, rightChild = sibling, activeNode
+		} else {
+			leftChild, rightChild = activeNode, sibling
+		}
+		activeNode = GetSha256Parent(leftChild, rightChild)
```

Comment:
if you choose to use a delegate for the parent calculation in the Merkle tree you must follow this pattern here as well